Update all eventlog entries to be full JSON objects

chu11 opened the pull request:
Per RFC18 update, this changes all eventlog entries to be full JSON objects.  This is the second part of #2106.

Note that there is no refactoring here, just the basics to get flux-core to be RFC18 compliant.  Part 3 will include a new `libeventlog` library to refactor things.

chu11 commented:
hmmm, haven't seen a hang in awhile.  Restarting builder.

```

No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

chu11 commented:
rebased

garlick commented:
So the context and event name max lengths went away in the RFC 18 updates.

How would you feel about adding the next PR to this one?   Since we create our release notes from the merge commits, it's nice if a PR completely handles one thing rather than being split over several PRs.  Is there a good reason to split it here?

(Apologies if we already discussed this and you made a good point that I've now forgotten!)

@grondo suggests if we agree that practice is what we want, we could add it to RFC 1.

chu11 commented:
> So the context and event name max lengths went away in the RFC 18 updates.

Yup.

> How would you feel about adding the next PR to this one? Since we create our release notes from the merge commits, it's nice if a PR completely handles one thing rather than being split over several PRs. Is there a good reason to split it here?

I just felt it was two different changes.  Converting to RFC18 felt like one change (i.e. nothing wrong with keeping the code the way it is and can keep the `flux_kvs_eventlog` API).  To me the inclusion of the `libeventlog` is a refactoring / cleanup that is separate from this one.

I can go either way on this.

garlick commented:
Sorry I was a bit terse above.  I meant the PR doesn't quite get the code in sync with RFC 18 because the old API enforces the old length limitations.  The argument for combining the PR's is that it accomplishes the entire "sync with RFC 18 changes" task in one go (one release notes entry), and some intermediate changes can be avoided (e.g. in a combined PR: add new libeventlog api + tests, convert users, drop old kvs/eventlog api + tests, thus avoiding updates the the old api and tests).

If thinking in the granularity of release notes, this seems like one part of a whole change, not an independent change.

Having said that, you've got this PR all prepared and the button is green. If you prefer, I'll hit the button and we can split hairs on the next topic :-)

chu11 commented:
> Sorry I was a bit terse above. I meant the PR doesn't quite get the code in sync with RFC 18 because the old API enforces the old length limitations. 

Ahhh.  I get it now.  I'll go ahead and push the bigger overall change.

chu11 commented:
re-pushed with the new `libeventlog` library and the removal of the old `kvs_eventlog`.

chu11 commented:
Lots of builders failed.  All tests pass but I clearly screwed up something in the Makefile.am that travis is expecting with my new unit tests.  Hmmm.

```
make  test_eventlog.t

  CC       test/test_eventlog_t-eventlog.o

  CCLD     test_eventlog.t

make  check-TESTS

           test_eventlog.t:  PASS: N=54  PASS=54  FAIL=0 SKIP=0 XPASS=0 XFAIL=0

ERROR: test_eventlog.t - missing test plan

tap-driver.sh: internal error getting exit status

tap-driver.sh: fatal: I/O or internal error

Makefile:1028: recipe for target 'test_eventlog.log' failed
```

chu11 commented:
hmmm, Possible b/c the logs have bad output?  I'll re-work and try again.

```
ok 70 - eventlog_entry_decode - decoded "{"timestamp":1.0,"name":"foo"}
" correctly
```

chu11 commented:
oh wait

```
           test_eventlog.t:  PASS: N=54  PASS=54  FAIL=0 SKIP=0 XPASS=0 XFAIL=0
```

I have > 54 unit tests.  Hmmmm, but little info to off of.

grondo commented:
Yeah, that's weird. The fact that the test got a `PASS` implies that the test script issued the TAP protocol `1..54`. Did you prematurely call `test_done`?

What does the `test-suite.log` output show?

grondo commented:
This branch fails for me with just plain `make check`, the `test_eventlog.t` is segfaulting after test 89:
```
ok 84 - eventlog_entry_create context="foo" fails with EINVAL
ok 85 - eventlog_entry_create context="["foo"]" fails with EINVAL
ok 86 - eventlog_entry_create context="{"data":"foo"}\n" fails with EINVAL
ok 87 - eventlog_entry_pack name=NULL fails with EINVAL
ok 88 - eventlog_entry_pack context="" fails with EINVAL
ok 89 - eventlog_entry_pack context="foo" fails with EINVAL
Segmentation fault (core dumped)
```
under gdb
```
Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7bce290 in ?? () from /usr/lib/x86_64-linux-gnu/libjansson.so.4
(gdb) where
#0  0x00007ffff7bce290 in ?? () from /usr/lib/x86_64-linux-gnu/libjansson.so.4
#1  0x00007ffff7bce785 in json_vpack_ex ()
   from /usr/lib/x86_64-linux-gnu/libjansson.so.4
#2  0x000055555555baeb in eventlog_entry_vpack (timestamp=1, 
    name=0x55555555c9c4 "foo", context_fmt=0x55555555ca26 "[\"foo\"]", 
    ap=ap@entry=0x7fffffffe2b0) at eventlog.c:276
#3  0x000055555555bc3f in eventlog_entry_pack (timestamp=<optimized out>, 
    name=<optimized out>, context_fmt=<optimized out>) at eventlog.c:298
#4  0x0000555555557338 in eventlog_entry_encoding_errors ()
    at test/eventlog.c:403
#5  0x0000555555555c55 in main (argc=<optimized out>, argv=<optimized out>)
    at test/eventlog.c:460
```
in case it is not reproducing for you

chu11 commented:
> Yeah, that's weird. The fact that the test got a PASS implies that the test script issued the TAP protocol 1..54. Did you prematurely call test_done?

Don't think so.  The main routine is pretty normal:

```
    plan (NO_PLAN);

    eventlog_entry_parsing ();
    eventlog_decoding ();
    eventlog_decoding_errors ();
    eventlog_entry_decoding ();
    eventlog_entry_decoding_errors ();
    eventlog_entry_encoding ();
    eventlog_entry_encoding_errors ();

    done_testing ();
```

> What does the test-suite.log output show?

Now that you mention it, there is an odd thing in `test-suite.log`.  There are no errors for `test_eventlog.t`, but there are errors for `test_toml.t`.

```
PASS: test_toml
===============
ok 1 - ex1: parsed simple example
PASS: test_toml.t 1 - ex1: parsed simple example
ok 2 - ex1: located server table
PASS: test_toml.t 2 - ex1: located server table
<SNIP>
not ok 111 - utf8_to_ucs: 0x80 converted from 2-char UTF8 # TODO https://github.com/cktan/tomlc99/issues/7
XFAIL: test_toml.t 111 - utf8_to_ucs: 0x80 converted from 2-char UTF8 # TODO https://github.com/cktan/tomlc99/issues/7
#   Failed (TODO) test 'utf8_to_ucs: 0x80 converted from 2-char UTF8'
#   at test/toml.c line 401.
not ok 112 - utf8_to_ucs: 0x7ff converted from 2-char UTF8 # TODO https://github.com/cktan/tomlc99/issues/7
XFAIL: test_toml.t 112 - utf8_to_ucs: 0x7ff converted from 2-char UTF8 # TODO https://github.com/cktan/tomlc99/issues/7
#   Failed (TODO) test 'utf8_to_ucs: 0x7ff converted from 2-char UTF8'
#   at test/toml.c line 403.
<SNIP - more errors>
```

Note that it indicates `PASS` for these tests, but there are clearly errors.

chu11 commented:
> This branch fails for me with just plain make check, the test_eventlog.t is segfaulting after test 89:

Thanks.  I'd been concentrating on mem corruption around test 55, but perhaps it's way earlier given you're seeing it at 90.

grondo commented:
Here's valgrind output from my system in case it helps:
```
==32145== Invalid read of size 8
==32145==    at 0x4E43290: ??? (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x4E43784: json_vpack_ex (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x10FAEA: eventlog_entry_vpack (eventlog.c:276)
==32145==    by 0x10FC3E: eventlog_entry_pack (eventlog.c:298)
==32145==    by 0x10B337: eventlog_entry_encoding_errors (eventlog.c:403)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Address 0x544e748 is 8 bytes inside a block of size 53 free'd
==32145==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B8C6: vok_at_loc (tap.c:85)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B311: eventlog_entry_encoding_errors (eventlog.c:398)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Block was alloc'd at
==32145==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B779: vstrdupf (tap.c:29)
==32145==    by 0x10B824: vok_at_loc (tap.c:61)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B311: eventlog_entry_encoding_errors (eventlog.c:398)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145== 
==32145== Invalid read of size 8
==32145==    at 0x4E432A5: ??? (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x4E43784: json_vpack_ex (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x10FAEA: eventlog_entry_vpack (eventlog.c:276)
==32145==    by 0x10FC3E: eventlog_entry_pack (eventlog.c:298)
==32145==    by 0x10B337: eventlog_entry_encoding_errors (eventlog.c:403)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Address 0x544e748 is 8 bytes inside a block of size 53 free'd
==32145==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B8C6: vok_at_loc (tap.c:85)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B311: eventlog_entry_encoding_errors (eventlog.c:398)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Block was alloc'd at
==32145==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B779: vstrdupf (tap.c:29)
==32145==    by 0x10B824: vok_at_loc (tap.c:61)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B311: eventlog_entry_encoding_errors (eventlog.c:398)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145== 
ok 90 - eventlog_entry_pack context="["foo"]" fails with EINVAL
ok 91 - eventlog_entry_pack context="{"data":"foo"}" fails with EINVAL
ok 92 - eventlog_entry_pack context="{"data":"foo"}\n" fails with EINVAL
ok 93 - eventlog_entry_vpack name=NULL fails with EINVAL
ok 94 - eventlog_entry_vpack context="" fails with EINVAL
ok 95 - eventlog_entry_vpack context="foo" fails with EINVAL
==32145== Invalid read of size 8
==32145==    at 0x4E43290: ??? (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x4E43784: json_vpack_ex (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x10FAEA: eventlog_entry_vpack (eventlog.c:276)
==32145==    by 0x10ABAE: test_eventlog_entry_vpack (eventlog.c:288)
==32145==    by 0x10B507: eventlog_entry_encoding_errors (eventlog.c:434)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Address 0x544ed58 is 8 bytes inside a block of size 54 free'd
==32145==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B8C6: vok_at_loc (tap.c:85)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B4E1: eventlog_entry_encoding_errors (eventlog.c:429)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Block was alloc'd at
==32145==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B779: vstrdupf (tap.c:29)
==32145==    by 0x10B824: vok_at_loc (tap.c:61)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B4E1: eventlog_entry_encoding_errors (eventlog.c:429)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145== 
==32145== Invalid read of size 8
==32145==    at 0x4E432A5: ??? (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x4E43784: json_vpack_ex (in /usr/lib/x86_64-linux-gnu/libjansson.so.4.11.0)
==32145==    by 0x10FAEA: eventlog_entry_vpack (eventlog.c:276)
==32145==    by 0x10ABAE: test_eventlog_entry_vpack (eventlog.c:288)
==32145==    by 0x10B507: eventlog_entry_encoding_errors (eventlog.c:434)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Address 0x544ed58 is 8 bytes inside a block of size 54 free'd
==32145==    at 0x4C30D3B: free (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B8C6: vok_at_loc (tap.c:85)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B4E1: eventlog_entry_encoding_errors (eventlog.c:429)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145==  Block was alloc'd at
==32145==    at 0x4C2FB0F: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==32145==    by 0x10B779: vstrdupf (tap.c:29)
==32145==    by 0x10B824: vok_at_loc (tap.c:61)
==32145==    by 0x10BAE6: ok_at_loc (tap.c:93)
==32145==    by 0x10B4E1: eventlog_entry_encoding_errors (eventlog.c:429)
==32145==    by 0x109C54: main (eventlog.c:460)
==32145== 
ok 96 - eventlog_entry_vpack context="["foo"]" fails with EINVAL
ok 97 - eventlog_entry_vpack context="{"data":"foo"}" fails with EINVAL
ok 98 - eventlog_entry_vpack context="{"data":"foo"}\n" fails with EINVAL
1..98
==32145== 
==32145== HEAP SUMMARY:
==32145==     in use at exit: 0 bytes in 0 blocks
==32145==   total heap usage: 689 allocs, 689 frees, 152,099 bytes allocated
==32145== 
==32145== All heap blocks were freed -- no leaks are possible
==32145== 
==32145== For counts of detected and suppressed errors, rerun with: -v
==32145== ERROR SUMMARY: 8 errors from 4 contexts (suppressed: 0 from 0)
grondo@asp:~/git/flux-core.git/src/common/libeventlog$ 
```

chu11 commented:
Hmmm.  What version of jansson is on your machine @grondo?  I'm wondering if my tests like:

```
    errno = 0;
    ok (eventlog_entry_pack (1., "foo", "[\"foo\"]") == NULL
        && errno == EINVAL,
        "eventlog_entry_pack context=\"[\"foo\"]\" fails with EINVAL");

```

are creating problems b/c the `o` in `foo` is being mistaken for a format specifier.   The test was supposed to be a simple "pass in an json array" test leading to EINVAL.

grondo commented:
> What version of jansson is on your machine @grondo? 

Ubuntu 18.04:
```
libjansson-dev:amd64	2.11-1
```

This matches what we've got in the `bionic` docker env:
```
$ docker run fluxrm/testenv:bionic dpkg-query -W libjansson-dev     
libjansson-dev:amd64    2.11-1
```

while CentOS 7 has 2.10. Maybe that is why the centos builds seem to succeed?

```
$ docker run fluxrm/testenv:centos7 rpm -q jansson-devel                                                                        
jansson-devel-2.10-1.el7.x86_64
```

chu11 commented:
> while CentOS 7 has 2.10. Maybe that is why the centos builds seem to succeed?

Yeah.  It appears my guess that the `o` in `foo` was being confused as a format specifier was the issue.  Thanks for the help!

One builder hit a valgrind issue, but no info beyond just this.  Hmmm.

```
==1668== 
==1668== HEAP SUMMARY:
==1668==     in use at exit: 23,615 bytes in 57 blocks
==1668==   total heap usage: 28,349 allocs, 28,292 frees, 19,706,125 bytes allocated
==1668== 
==1668== LEAK SUMMARY:
==1668==    definitely lost: 0 bytes in 0 blocks
==1668==    indirectly lost: 0 bytes in 0 blocks
==1668==      possibly lost: 0 bytes in 0 blocks
==1668==    still reachable: 23,615 bytes in 57 blocks
==1668==         suppressed: 0 bytes in 0 blocks
==1668== Reachable blocks (those to which a pointer was found) are not shown.
==1668== To see them, rerun with: --leak-check=full --show-leak-kinds=all
```

chu11 commented:
re-pushed, for whatever reason, jansson 2.10 was able to correctly tell that a format like `[\"foo\"]` was invalid and return EINVAL.  But in jansson 2.11, I believe the `f` and `o` are interpreted as format specifiers and badness occurs.

garlick commented on src/common/libeventlog/eventlog.c:
```diff
@@ -0,0 +1,331 @@
+    if (name)
+        (*name) = n;
+
+    if (!json_unpack (entry, "{ s:o }", "context", &c)) {
```
There probably needs to be a check that 'o' is a JSON object and not an array.

garlick commented on src/common/libeventlog/eventlog.c:
```diff
@@ -0,0 +1,331 @@
+    }
+
+    if (!(copy = strdup (s)))
+        goto enomem;
```
strdup sets ENOMEM so just return NULL?

garlick commented on src/common/libeventlog/eventlog.c:
```diff
@@ -0,0 +1,331 @@
+    return true;
+}
+
+static json_t *eventlog_entry_decode_wrapper (const char *entry,
```
Different name?  It's not a wrapper for event_decode(), it's the other way around right?

garlick commented on src/common/libeventlog/eventlog.c:
```diff
@@ -0,0 +1,331 @@
+    if (timestamp < 0.
+        || !name
+        || !strlen (name)
+        || strchr (name, ' ') != NULL
```
RFC 18 doesn't actually impose any restrictions on what is in the "name" string anymore, so prob OK to skip these checks for white space.  (note embedded newlines are escaped in JSON strings and cannot be confused with the event end of line)

garlick commented on src/common/libeventlog/eventlog.c:
```diff
@@ -0,0 +1,331 @@
+    int save_errno;
+
+    if (context) {
+        if (strchr (context, '\n')) {
```
embedded newline check not needed.

garlick commented on src/common/libeventlog/eventlog.c:
```diff
@@ -0,0 +1,331 @@
+    int save_errno;
+
+    if (context_fmt) {
+        if (strchr (context_fmt, '\n')) {
```
check for embedded newline not needed

garlick commented on src/common/libeventlog/eventlog.c:
```diff
@@ -0,0 +1,331 @@
+    if (!(s = json_dumps (entry, JSON_COMPACT)))
+        goto enomem;
+    if (asprintf (&rv, "%s\n", s) < 0)
+        goto enomem;
```
asprintf sets errno so might want to not re-set it here.

chu11 commented:
> I think you can squash out of existence the changes to libkvs/eventlog.[ch] and its test by reordering commits?

Ahhh yes, now that I've put the two PRs into one, that initial change makes less sense.  I'll see how I can piece things together into a more sensible commit series.

chu11 commented:
just re-pushed with suggested fixes above and squashed changes into a smaller set of commits.  Added new tests such as `e = eventlog_entry_pack (1., "foo", "{ s:s }", "data", "foo\n");` to ensure newlines in contexts work as expected.

garlick commented:
LGTM!  Just needs a rebase.

chu11 commented:
rebased

garlick commented:
Awesome!
